File: BasicAlgWithJava/剑指offerWithJava/src/Sort/HeapSort.java
package Sort;

import java.util.Arrays;

import static DataStructure.swap.swap;

@SuppressWarnings({"all"})
public class HeapSort {
    public static void heapSort(int[] nums) {
        if (nums == null || nums.length < 2) {
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            heapinsert(nums, i);
        }
        int size = nums.length;
        while (size > 0) {
            swap(nums, 0, --size);
            heapify(nums, 0, size);
        }


    }

    public static void heapinsert(int[] nums, int index) {
        while (nums[index] > nums[(index - 1) / 2]) {
            swap(nums, index, (index - 1) / 2);
            index = (index - 1) / 2;

        }
    }

    public static void heapify(int[] nums, int index, int size) {
        int left = index * 2 + 1;
        while (left < size) {
            int largest = (left + 1 < size) && (nums[left + 1] > nums[left]) ? left + 1 : left;
            largest = nums[largest] > nums[index] ? largest : index;
            if (index == largest) {
                break;
            }
            swap(nums, index, largest);
            index = largest;
            left = 2 * index + 1;
        }
    }

    public static void main(String[] args) {
        int[] nums = {1, 0, 9, 8, 7, 2, 3, 6, 5, 4, 13, 11, 12};
        heapSort(nums);
        System.out.println("Arrays.toString(nums) = " + Arrays.toString(nums));
        System.out.println("Arrays.stream(nums).toArray() = " + Arrays.stream(nums).toArray());
        System.out.println("nums = " + nums);

    }


}

File: LC-Go/Sorts/HeapSort.go
package Sorts

func heapSort(nums []int) []int {
	if len(nums) == 0 {
		return nil
	}
	for i := 0; i < len(nums); i++ {
		heapInsert(nums, i)
	}
	size := len(nums)
	for size > 0 {
		size--
		Swap(nums, 0, size)
		heapIfy(nums, 0, size)
	}
	return nums
}
func heapIfy(nums []int, index int, size int) {
	left := index*2 + 1
	for left < size {
		var largest int
		if left+1 < size && nums[left+1] > nums[left] {
			largest = left + 1
		} else {
			largest = left
		}
		if nums[largest] < nums[index] {
			break
		}
		Swap(nums, index, largest)
		index = largest
		left = 2*index + 1
	}
}

func heapInsert(nums []int, index int) {
	for nums[index] > nums[(index-1)/2] {
		Swap(nums, index, (index-1)/2)
		index = (index - 1) / 2
	}
}

func Swap(nums []int, j int, i int) {
	temp := nums[j]
	nums[j] = nums[i]
	nums[i] = temp
}
